src/codegen/encoding/scale_encoding.rs

```diff
@@ -1,66 +1,200 @@
+        },
+    );
+    cfg.add(vartab, Instr::Branch { block: done });
+
+    cfg.set_basic_block(medium);
+    if let (Some(buffer), Some(offset)) = (buffer, offset) {
+        let mul2 = Expression::BitwiseOr(Codegen, Uint(32), mul.clone().into(), one.into());
+        cfg.add(
+            vartab,
+            Instr::WriteBuffer {
+                buf: buffer.clone(),
+                offset: offset.clone(),
+                value: mul2,
+            },
+        );
+    }
+    let two = Expression::NumberLiteral(Codegen, Uint(32), 2.into());
+    cfg.add(
+        vartab,
+        Instr::Set {
+            loc: Codegen,
+            res: size_variable,
+            expr: two.clone(),
+        },
+    );
+    cfg.add(vartab, Instr::Branch { block: done });
+
+    cfg.set_basic_block(big);
+    if let (Some(buffer), Some(offset)) = (buffer, offset) {
+        let mul2 = Expression::BitwiseOr(Codegen, Uint(32), mul.into(), two.into());
+        cfg.add(
+            vartab,
+            Instr::WriteBuffer {
+                buf: buffer.clone(),
+                offset: offset.clone(),
+                value: mul2,
+            },
+        );
+    }
+    cfg.add(
+        vartab,
+        Instr::Set {
+            loc: Codegen,
+            res: size_variable,
+            expr: Expression::NumberLiteral(Codegen, Uint(32), 4.into()),
+        },
+    );
+    cfg.add(vartab, Instr::Branch { block: done });
+
+    cfg.set_basic_block(done);
+    cfg.set_phis(done, vartab.pop_dirty_tracker());
+    Expression::Variable(Codegen, Uint(32), size_variable)
+}
+
+impl AbiEncoding for ScaleEncoding {
+    fn size_width(
+        &self,
+        size: &Expression,
+        vartab: &mut Vartable,
+        cfg: &mut ControlFlowGraph,
+    ) -> Expression {
+        encode_compact(size, None, None, vartab, cfg)
+    }
 
-        let encoded_expr = Expression::Variable(*loc, Type::DynamicBytes, encoded_buffer);
-        let buffer_len = Expression::Builtin(
-            *loc,
-            vec![Type::Uint(32)],
-            Builtin::ArrayLength,
-            vec![encoded_expr.clone()],
+    fn encode_external_function(
+        &mut self,
+        expr: &Expression,
+        buffer: &Expression,
+        offset: &Expression,
+        ns: &Namespace,
+        vartab: &mut Vartable,
+        cfg: &mut ControlFlowGraph,
+    ) -> Expression {
+        let addr_len = ns.address_length.into();
+        let address = expr.external_function_address();
+        let size = self.encode_direct(&address, buffer, offset, vartab, cfg, addr_len);
+        let offset = Expression::Add(
+            Codegen,
+            Uint(32),
+            false,
+            offset.clone().into(),
+            size.clone().into(),
         );
+        let selector = expr.external_function_selector();
+        let selector_size = self.encode_direct(&selector, buffer, &offset, vartab, cfg, 4.into());
+        Expression::Add(Codegen, Uint(32), false, size.into(), selector_size.into())
```

Comment:
The selector and the address have sizes that are known during compile time. There is no need to create an addition here.
In addition, this function does exactly the same thing you do on bosh encoding. I believe you can have a single implementation for both cases.

Reply:
> The selector and the address have sizes that are known during compile time. There is no need to create an addition here.

Well I don't think adding vs. loading a value make a huge difference and this code most likely never ends up inside hot code paths. But I can change it

> In addition, this function does exactly the same thing you do on bosh encoding

For solana the selector comes first